spec/client/ics-012-near-client/README.md

```diff
@@ -0,0 +1,550 @@
+    switch typeof(clientMessage) {
+      case Header:
+        verifyHeader(clientMessage)
+    }
+}
+```
+
+Verify validity of regular update to the NEAR client
+
+```typescript
+function (Header) getEpochId(): CryptoHash {
+    return self.lightClientBlock.innerLite.epochId
+}
+
+function (Header) getNextEpochId(): CryptoHash {
+    return self.lightClientBlock.innerLite.nextEpochId
+}
+
+function (ConsensusState) getBlockProducersOf(epochId: CryptoHash): List<ValidatorStakeView> {
+    if epochId === self.header.getEpochId() {
+        return self.currentBps
+    } else if epochId === self.header.getNextEpochId() {
+        return self.header.lightClientBlock.nextBps
+    } else {
+        return null
+    }
+}
+
+function verifyHeader(header: Header) {
+    clientState = get("clients/{clientMessage.identifier}/clientState")
+
+    latestHeader = clientState.getLatestHeader()
+    approvalMessage = header.lightClientBlock.approvalMessage()
+
+    // (1) The height of the block is higher than the height of the current head.
+    assert(clientState.latestHeight < header.getHeight())
+
+    // (2) The epoch of the block is equal to the epochId or nextEpochId
+    //     known for the current head.
+    assert(header.getEpochId() in
+        [latestHeader.getEpochId(), latestHeader.getNextEpochId()])
+
+    // (3) If the epoch of the block is equal to the nextEpochId of the head,
+    //     then nextBps is not null.
+    assert(not(header.getEpochId() == latestHeader.getNextEpochId()
+        && header.lightClientBlock.nextBps === null))
+
+    // (4) approvalsAfterNext contain valid signatures on approvalMessage
+    //     from the block producers of the corresponding epoch.
+    // (5) The signatures present in approvalsAfterNext correspond to
+    //     more than 2/3 of the total stake.
+    totalStake = 0
+    approvedStake = 0
+
+    epochBlockProducers = clientState.getBlockProducersOf(header.getEpochId())
+    for maybeSignature, blockProducer in
+      zip(header.lightClientBlock.approvalsAfterNext, epochBlockProducers) {
+        totalStake += blockProducer.stake
+
+        if maybeSignature === null {
+            continue
+        }
+
+        approvedStake += blockProducer.stake
+
+        assert(verify_signature(
+            public_key: blockProducer.public_key,
+            signature: maybeSignature,
+            message: approvalMessage
+        ))
+    }
+
+    assert(approvedStake * 3 > totalStake * 2)
+
+    // (6) If nextBps is not none, hash(borsh(nextBps)) corresponds to the nextBpHash in innerLite
+    if header.lightClientBlock.nextBps !== null {
+        assert(hash(borsh(header.lightClientBlock.nextBps))
+            === header.lightClientBlock.innerLite.nextBpHash)
+    }
+
+    // (7) Check the prevStateRoot is the root of merklized prevStateRootOfChunks
+    assert(header.lightClientBlock.innerLite.prevStateRoot
+        === merklize(header.prevStateRootOfChunks).root)
```

Comment:
In implementation, this function could return more data (for example, the paths of all nodes in the tree) rather than only the root. So I add a description for this function at line 39. I think it's clear now.